# Modul 3 Sains Data

Kembali ke [Sains Data](./saindat2024genap.qmd)

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Install `scikit-learn` dengan:

`!pip install scikit-learn`

Lalu import `sklearn`:

In [7]:
import sklearn

## Import Dataset

Untuk praktikum kali ini, kita akan menggunakan dataset "California Housing Prices" (housing.csv) yang bisa didownload dari salah satu sumber berikut:

* [Direct link (langsung dari GitHub Pages ini)](./housing.csv)

* Kaggle: <https://www.kaggle.com/datasets/camnugent/california-housing-prices>

Kemudian, baca sebagai dataframe:

In [3]:
df = pd.read_csv("./housing.csv")

Mari kita lihat isinya:

In [4]:
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

Ada satu data kategorik, yaitu `ocean_proximity`. Mari kita liat jenis-jenisnya (kategorinya):

In [69]:
df["ocean_proximity"].value_counts()

ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64

Apakah ada *missing value*?

In [5]:
df.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [6]:
df[df["total_bedrooms"].isna()]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
290      -122.16     37.77                47.0       1256.0             NaN   
341      -122.17     37.75                38.0        992.0             NaN   
538      -122.28     37.78                29.0       5154.0             NaN   
563      -122.24     37.75                45.0        891.0             NaN   
696      -122.10     37.69                41.0        746.0             NaN   
...          ...       ...                 ...          ...             ...   
20267    -119.19     34.20                18.0       3620.0             NaN   
20268    -119.18     34.19                19.0       2393.0             NaN   
20372    -118.88     34.17                15.0       4260.0             NaN   
20460    -118.75     34.29                17.0       5512.0             NaN   
20484    -118.72     34.28                17.0       3051.0             NaN   

       population  households  median_income  median_house_value  \
290         570.0       218.0         4.3750            161900.0   
341         732.0       259.0         1.6196             85100.0   
538        3741.0      1273.0         2.5762            173400.0   
563         384.0       146.0         4.9489            247100.0   
696         387.0       161.0         3.9063            178400.0   
...           ...         ...            ...                 ...   
20267      3171.0       779.0         3.3409            220500.0   
20268      1938.0       762.0         1.6953            167400.0   
20372      1701.0       669.0         5.1033            410700.0   
20460      2734.0       814.0         6.6073            258100.0   
20484      1705.0       495.0         5.7376            218600.0   

      ocean_proximity  
290          NEAR BAY  
341          NEAR BAY  
538          NEAR BAY  
563          NEAR BAY  
696          NEAR BAY  
...               ...  
20267      NEAR OCEAN  
20268      NEAR OCEAN  
20372       <1H OCEAN  
20460       <1H OCEAN  
20484       <1H OCEAN  

[207 rows x 10 columns]

Perhatikan bahwa tipe datanya adalah `int64` atau bilangan bulat.

Dari 20640 baris, ada satu kolom/fitur (`total_bedrooms`) dengan 207 missing value.

Secara umum, ada dua cara untuk menangani *missing value*:

1. Menghapus baris-baris yang memiliki *missing value*, dengan `df.dropna()`

2. Melakukan metode imputasi

Karena banyaknya *missing value* relatif sedikit, sebenarnya tidak masalah apabila baris-baris tersebut cukup dihapus saja. Namun, kita akan mempelajari metode imputasi.

## Encoding Data Kategorik

Sebelum kita membahas metode imputasi, kita akan membahas tentang melakukan "encoding" untuk data kategorik.

Banyak metode sains data / *machine learning* yang hanya bisa digunakan dengan data numerik. Oleh karena itu, data kategorik perlu diubah terlebih dahulu menjadi data numerik, melakukan yang namanya *categorical data encoding*

Metode yang sering digunakan adalah *one hot encoding*. Misalnya ada satu fitur kategorik dengan $n$ kemungkinan data, bernama $D_i$ untuk $i = 1, 2, \dots, n$. Maka fitur tersebut diganti dengan $n$ kolom baru, misal bernama $K_i$ untuk $i = 1, 2, \dots, n$, di mana pada kolom ke-i, isinya adalah

* $1$, apabila data aslinya pada baris tersebut adalah $D_i$
* $0$ apabila bukan $D_i$

In [70]:
from sklearn.preprocessing import OneHotEncoder

In [72]:
encoder = OneHotEncoder()

In [83]:
hasil_onehot = encoder.fit_transform(df[["ocean_proximity"]])

In [87]:
print(encoder.categories_)

[array(['<1H OCEAN', 'INLAND', 'ISLAND', 'NEAR BAY', 'NEAR OCEAN'],
      dtype=object)]


In [89]:
print(encoder.categories_[0])

['<1H OCEAN' 'INLAND' 'ISLAND' 'NEAR BAY' 'NEAR OCEAN']


In [142]:
kolom_encoding = list(encoder.categories_[0])

In [143]:
print(kolom_encoding)

['<1H OCEAN', 'INLAND', 'ISLAND', 'NEAR BAY', 'NEAR OCEAN']


In [84]:
onehot_array = hasil_onehot.toarray()

In [85]:
print(onehot_array)

[[0. 0. 0. 1. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 1. 0.]
 ...
 [0. 1. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 1. 0. 0. 0.]]


In [144]:
onehot_df = pd.DataFrame(onehot_array, columns=kolom_encoding)

In [145]:
onehot_df

<1H OCEAN  INLAND  ISLAND  NEAR BAY  NEAR OCEAN
0            0.0     0.0     0.0       1.0         0.0
1            0.0     0.0     0.0       1.0         0.0
2            0.0     0.0     0.0       1.0         0.0
3            0.0     0.0     0.0       1.0         0.0
4            0.0     0.0     0.0       1.0         0.0
...          ...     ...     ...       ...         ...
20635        0.0     1.0     0.0       0.0         0.0
20636        0.0     1.0     0.0       0.0         0.0
20637        0.0     1.0     0.0       0.0         0.0
20638        0.0     1.0     0.0       0.0         0.0
20639        0.0     1.0     0.0       0.0         0.0

[20640 rows x 5 columns]

In [92]:
df = pd.concat([df, onehot_df], axis=1)

In [93]:
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  <1H OCEAN  INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
0            NEAR BAY        0.0     0.0     0.0       1.0         0.0  
1            NEAR BAY        0.0     0.0     0.0       1.0         0.0  
2            NEAR BAY        0.0     0.0     0.0       1.0         0.0  
3            NEAR BAY        0.0     0.0     0.0       1.0         0.0  
4            NEAR BAY        0.0     0.0     0.0       1.0         0.0  
...               ...        ...     ...     ...       ...         ...  
20635          INLAND        0.0     1.0     0.0       0.0         0.0  
20636          INLAND        0.0     1.0     0.0       0.0         0.0  
20637          INLAND        0.0     1.0     0.0       0.0         0.0  
20638          INLAND        0.0     1.0     0.0       0.0         0.0  
20639          INLAND        0.0     1.0     0.0       0.0         0.0  

[20640 rows x 15 columns]

In [150]:
df = df.drop(["ocean_proximity"], axis=1)

In [151]:
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  <1H OCEAN  \
0           322.0       126.0         8.3252            452600.0        0.0   
1          2401.0      1138.0         8.3014            358500.0        0.0   
2           496.0       177.0         7.2574            352100.0        0.0   
3           558.0       219.0         5.6431            341300.0        0.0   
4           565.0       259.0         3.8462            342200.0        0.0   
...           ...         ...            ...                 ...        ...   
20635       845.0       330.0         1.5603             78100.0        0.0   
20636       356.0       114.0         2.5568             77100.0        0.0   
20637      1007.0       433.0         1.7000             92300.0        0.0   
20638       741.0       349.0         1.8672             84700.0        0.0   
20639      1387.0       530.0         2.3886             89400.0        0.0   

       INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
0         0.0     0.0       1.0         0.0  
1         0.0     0.0       1.0         0.0  
2         0.0     0.0       1.0         0.0  
3         0.0     0.0       1.0         0.0  
4         0.0     0.0       1.0         0.0  
...       ...     ...       ...         ...  
20635     1.0     0.0       0.0         0.0  
20636     1.0     0.0       0.0         0.0  
20637     1.0     0.0       0.0         0.0  
20638     1.0     0.0       0.0         0.0  
20639     1.0     0.0       0.0         0.0  

[20640 rows x 14 columns]

## Metode Imputasi

### Median

In [152]:
df_fill_median = df.copy()

In [153]:
df["total_bedrooms"].median()

435.0

In [154]:
bedrooms_median = df["total_bedrooms"].median()
print(bedrooms_median)

435.0


In [155]:
df_fill_median["total_bedrooms"] = df_fill_median["total_bedrooms"].fillna(bedrooms_median)

In [156]:
df_fill_median.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
<1H OCEAN             0
INLAND                0
ISLAND                0
NEAR BAY              0
NEAR OCEAN            0
dtype: int64

Cara lain, menggunakan scikit-learn:

In [157]:
df_fill_median2 = df.copy()

In [158]:
from sklearn.impute import SimpleImputer

In [159]:
median_imputer = SimpleImputer(
    missing_values=np.nan, strategy='median'
)

In [160]:
df_fill_median2[["total_bedrooms"]] = median_imputer.fit_transform(
    df_fill_median2[["total_bedrooms"]]
)

In [161]:
df_fill_median2.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
<1H OCEAN             0
INLAND                0
ISLAND                0
NEAR BAY              0
NEAR OCEAN            0
dtype: int64

### Modus

In [162]:
df_fill_mode = df.copy()

In [163]:
df["total_bedrooms"].mode()

0    280.0
Name: total_bedrooms, dtype: float64

In [164]:
df["total_bedrooms"].mode()[0]

280.0

In [165]:
bedrooms_mode = df["total_bedrooms"].mode()[0]
print(bedrooms_mode)

280.0


In [166]:
df_fill_mode["total_bedrooms"] = df_fill_mode["total_bedrooms"].fillna(bedrooms_mode)

In [167]:
df_fill_mode.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
<1H OCEAN             0
INLAND                0
ISLAND                0
NEAR BAY              0
NEAR OCEAN            0
dtype: int64

Cara lain, menggunakan scikit-learn:

In [168]:
df_fill_mode2 = df.copy()

In [169]:
mode_imputer = SimpleImputer(
    missing_values=np.nan, strategy='most_frequent'
)

In [170]:
df_fill_mode2[["total_bedrooms"]] = mode_imputer.fit_transform(
    df_fill_mode2[["total_bedrooms"]]
)

In [171]:
df_fill_mode2.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
<1H OCEAN             0
INLAND                0
ISLAND                0
NEAR BAY              0
NEAR OCEAN            0
dtype: int64

### Mean (rata-rata)

In [172]:
df_fill_mean = df.copy()

In [173]:
df_fill_mean["total_bedrooms"].mean()

537.8705525375618

In [174]:
np.round(df_fill_mean["total_bedrooms"].mean())

538.0

In [175]:
bedrooms_mean = np.round(df_fill_mean["total_bedrooms"].mean())
print(bedrooms_mean)

538.0


In [176]:
df_fill_mean["total_bedrooms"] = df_fill_mean["total_bedrooms"].fillna(bedrooms_mean)

In [177]:
df_fill_mean.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
<1H OCEAN             0
INLAND                0
ISLAND                0
NEAR BAY              0
NEAR OCEAN            0
dtype: int64

Cara lain, menggunakan scikit-learn:

In [178]:
df_fill_mean2 = df.copy()

In [179]:
mean_imputer = SimpleImputer(
    missing_values=np.nan, strategy='mean'
)

In [180]:
df_fill_mean2[["total_bedrooms"]] = mean_imputer.fit_transform(
    df_fill_mean2[["total_bedrooms"]]
)

In [181]:
df_fill_mean2.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
<1H OCEAN             0
INLAND                0
ISLAND                0
NEAR BAY              0
NEAR OCEAN            0
dtype: int64

### KNNImputer

In [182]:
from sklearn.impute import KNNImputer

In [183]:
knn_imputer = KNNImputer(n_neighbors=3)

In [184]:
df_fill_knn = df.copy()

KNN Imputer memerlukan kolom-kolom lainnya sebagai acuan, dan hanya bisa bekerja dengan data numerik. Sehingga, kita perlu mem-filter terlebih dahulu kolom-kolom numerik dari dataset kita.

In [185]:
df_fill_knn.select_dtypes(include='number')

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  <1H OCEAN  \
0           322.0       126.0         8.3252            452600.0        0.0   
1          2401.0      1138.0         8.3014            358500.0        0.0   
2           496.0       177.0         7.2574            352100.0        0.0   
3           558.0       219.0         5.6431            341300.0        0.0   
4           565.0       259.0         3.8462            342200.0        0.0   
...           ...         ...            ...                 ...        ...   
20635       845.0       330.0         1.5603             78100.0        0.0   
20636       356.0       114.0         2.5568             77100.0        0.0   
20637      1007.0       433.0         1.7000             92300.0        0.0   
20638       741.0       349.0         1.8672             84700.0        0.0   
20639      1387.0       530.0         2.3886             89400.0        0.0   

       INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
0         0.0     0.0       1.0         0.0  
1         0.0     0.0       1.0         0.0  
2         0.0     0.0       1.0         0.0  
3         0.0     0.0       1.0         0.0  
4         0.0     0.0       1.0         0.0  
...       ...     ...       ...         ...  
20635     1.0     0.0       0.0         0.0  
20636     1.0     0.0       0.0         0.0  
20637     1.0     0.0       0.0         0.0  
20638     1.0     0.0       0.0         0.0  
20639     1.0     0.0       0.0         0.0  

[20640 rows x 14 columns]

In [186]:
df_fill_knn.select_dtypes(include='number').columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', '<1H OCEAN', 'INLAND', 'ISLAND', 'NEAR BAY',
       'NEAR OCEAN'],
      dtype='object')

In [187]:
num_col = df_fill_knn.select_dtypes(include='number').columns

In [188]:
df_fill_knn[num_col] = knn_imputer.fit_transform(df_fill_knn[num_col])

In [189]:
df_fill_knn.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
<1H OCEAN             0
INLAND                0
ISLAND                0
NEAR BAY              0
NEAR OCEAN            0
dtype: int64

## Perbandingan Metode Imputasi

Kita bisa membandingkan beberapa metode imputasi (dan memilih yang mana yang terbaik) dengan langkah-langkah berikut.

1. Use a sample of your own dataset that does not contain any missing data (will serve as ground truth).
2. Introduce increasing proportions of missing data at random (e.g. 5–50 % in 5 % increments).
3. Reconstruct the missing data using the various methods.
4. Compute the sum of squared errors between the reconstructed and the original data, for each method and each proportion of missing data.

Langkah pertama, kita perlu memperoleh sample dari dataset kita yang tidak mengandung *missing value*, yang bisa disebut *ground truth*. Cara termudah adalah dengan menghapus baris-baris yang memiliki *missing value* (biasanya dipilih lagi sample hanya sebagian baris, tapi di sini tidak kita lakukan):

In [190]:
df_ground_truth = df.dropna()

In [191]:
df_ground_truth

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  <1H OCEAN  \
0           322.0       126.0         8.3252            452600.0        0.0   
1          2401.0      1138.0         8.3014            358500.0        0.0   
2           496.0       177.0         7.2574            352100.0        0.0   
3           558.0       219.0         5.6431            341300.0        0.0   
4           565.0       259.0         3.8462            342200.0        0.0   
...           ...         ...            ...                 ...        ...   
20635       845.0       330.0         1.5603             78100.0        0.0   
20636       356.0       114.0         2.5568             77100.0        0.0   
20637      1007.0       433.0         1.7000             92300.0        0.0   
20638       741.0       349.0         1.8672             84700.0        0.0   
20639      1387.0       530.0         2.3886             89400.0        0.0   

       INLAND  ISLAND  NEAR BAY  NEAR OCEAN  
0         0.0     0.0       1.0         0.0  
1         0.0     0.0       1.0         0.0  
2         0.0     0.0       1.0         0.0  
3         0.0     0.0       1.0         0.0  
4         0.0     0.0       1.0         0.0  
...       ...     ...       ...         ...  
20635     1.0     0.0       0.0         0.0  
20636     1.0     0.0       0.0         0.0  
20637     1.0     0.0       0.0         0.0  
20638     1.0     0.0       0.0         0.0  
20639     1.0     0.0       0.0         0.0  

[20433 rows x 14 columns]

Selanjutnya, kita perlu membolong-bolongi dataset ini, agar sekian persen diisi *missing value*.

In [192]:
import random

In [193]:
def missing_value_generator(df_original, percentage):
    df_miss = df_original.copy()
    baris, kolom = df_miss.shape
    n_total = baris*kolom

    permutasi = list(range(n_total))
    random.shuffle(permutasi)

    n_pilih = int(percentage * n_total)
    pilihan = permutasi[0 : n_pilih]

    for p in pilihan:
        df_miss.iloc[ int(p/kolom), p%kolom ] = np.nan
    
    return df_miss

In [194]:
df_miss_5 = missing_value_generator(df, 0.05)

In [195]:
df_miss_5.isna().sum()

longitude             1058
latitude              1016
housing_median_age    1054
total_rooms           1014
total_bedrooms        1219
population            1068
households            1036
median_income         1054
median_house_value    1047
<1H OCEAN             1010
INLAND                1011
ISLAND                1021
NEAR BAY               998
NEAR OCEAN            1042
dtype: int64

## Regresi Linier